eth2/utils/ssz/src/lib.rs

```diff
@@ -46,6 +46,8 @@ pub use encode::{Encode, SszEncoder};
 /// The number of bytes used to represent an offset.
 pub const BYTES_PER_LENGTH_OFFSET: usize = 4;
 /// The maximum value that can be represented using `BYTES_PER_LENGTH_OFFSET`.
+// allows overflow on 32-bit target
+#[allow(const_err)]
```

Comment:
Rather than relying on the `--release` mode overflow behaviour and silencing this warning, I think it would be clearer to rewrite this constant in terms of [`overflowing_pow`](https://doc.rust-lang.org/std/primitive.usize.html#method.overflowing_pow) and [`wrapping_sub`](https://doc.rust-lang.org/std/primitive.usize.html#method.wrapping_sub), like:

```rust
2usize.overflowing_pow(BYTES_PER_LENGTH_OFFSET as u32 * 8).0.wrapping_sub(1)
```

Otherwise it will panic in `debug` mode.

Aside: Interestingly, there doesn't seem to be a version of `<<`/`shl` that overflows the way we want. On 64-bit, `1usize.overflowing_shl(64).0` and `1usize.wrapping_shl(64)` will both give 1 rather than 0, because they mask out the bad bits before the shift.

Reply:
Actually, I suspect `(1 << 32) - 1` will yield 0 on 32-bit platforms, because `<<` behaves like the `shl` functions (at least on x86)

Reply:
Yeah this is interesting -- you're right, silencing the warning is not the right way forward. The only issue with `overflowing_pow` is it isn't a `const fn` and therefore cannot be calculated at compile time. It looks like `if` statements are not yet implemented based on the [RFC](https://github.com/rust-lang/rfcs/blob/master/text/0911-const-fn.md). Until that is resolved, I think there are only a few solutions:

* use the literal value of `2 ** 32 - 1`
* `std::u32::MAX`
* `0u32.wrapping_sub(1)`
* [lazy_static](https://docs.rs/lazy_static/1.3.0/lazy_static/) macro

Any preferences or do you have a different approach?

Reply:
Here's an option that maintains the relationship between `BYTES_PER_LENGTH_OFFSET` and `MAX_LENGTH_VALUE`:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e5ae81f366b2e6892cc23f74029b90bf

It also throws an error when `BYTES_PER_LENGTH_OFFSET` > number of bits in u32 (4), which is nice.

Reply:
I'm happy with Paul's solution and a `#[cfg()]` flag to choose between them. Alternatively, we could do the shifting op inside a `u64` where we know it will succeed, and then cast back to `usize` once we have the result, like:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4ffb6f67d899af585227adaca1031621

Reply:
@michaelsproul yours is more readable, but it doesn't give a compile error when `BYTES_PER_LENGTH_OFFSET` > number of bytes in u32 (4). Instead, it remains at `2^32 - 1`.

 I don't think this is a big deal though, happy for either solution. Would probably lean towards the uglier, "more safe" solution.

Reply:
Okay I just pushed this change -- I wasn't 100% clear if we were wanting to cast the `u32`/`u64` to a `usize` and do a `cfg` to pick the correct implementation per architecture, but that is what I went with. Let me know if I misunderstood what the consensus was!

Reply:
Yeah, looks good to me :)